### PR TITLE
bugfix lod cpu performance

### PR DESCRIPTION
**Pull request by Superjomn**
- The `Vector` is used to share data between CPU and GPU
- The is a mutex in `Vector`, results in serious performance issue in CPU mode
- so replace the `Vector` with STL `std::vector` in CPU compile
- still needs some through fix.

### CODE REVIEW

File: paddle/fluid/framework/lod_tensor.h [hunk 1]
```diff
@@ -55,7 +55,12 @@ namespace framework {
  *    0 2 4 7
  *    0 2 5 7 10 12 15 20
  */
+#if defined(PADDLE_WITH_CUDA)
```
Comment (luotao1) [inline paddle/fluid/framework/lod_tensor.h]: #ifdef PADDLE_WITH_CUDA

**Review by luotao1 — COMMENTED** [review]
Do you test how much speedup after this PR?

File: paddle/fluid/framework/lod_tensor.h (hunk 1 above)
Comment (Superjomn) [reply paddle/fluid/framework/lod_tensor.h]: This only affects NLP models, the NLP tested https://github.com/PaddlePaddle/Paddle/compare/develop...Superjomn:bugfix/lod-Vector2vector, and that saves a lot of performance.

Comment (Superjomn) [reply paddle/fluid/framework/lod_tensor.h]: https://gcc.gnu.org/onlinedocs/cpp/Defined.html

`#if defined` is equivalent to `#ifdef`

File: paddle/fluid/framework/mixed_vector.h [hunk 2]
```diff
@@ -26,6 +26,7 @@
 namespace paddle {
 namespace framework {
 
+#if defined(PADDLE_WITH_CUDA)
```
Comment (luotao1) [inline paddle/fluid/framework/mixed_vector.h:29]: `mixed_vector.h`没法统一了么？必须用宏隔开？

File: paddle/fluid/operators/adam_op.h [hunk 3]
```diff
@@ -293,11 +293,18 @@ class AdamOpKernel : public framework::OpKernel<T> {
       auto& grad_tensor = grad_merge.value();
       const T* grad_data = grad_tensor.template data<T>();
       int64_t* rows = nullptr;
+// When compiled without CUDA, the CUDAMutableData() interface should not be
```
Comment (luotao1) [inline paddle/fluid/operators/adam_op.h:296]: 只修改adam_op.h, target_assign_op.h, sequence2batch.h是因为性能还是编译问题？

**Review by luotao1 — APPROVED** [review]
LGTM

File: paddle/fluid/framework/mixed_vector.h (hunk 2 above)
Comment (Superjomn) [reply paddle/fluid/framework/mixed_vector.h:29]: 目前比较难，旧的 Vector 是 CPU<->GPU 的场景下用的，有一些接口跟 `std::vector` 不相同。 用宏彻底隔开 thrust 和 std::vector，短期会稳定一些，长期还需要多一些时间看看。

File: paddle/fluid/operators/adam_op.h (hunk 3 above)
Comment (Superjomn) [reply paddle/fluid/operators/adam_op.h:296]: 解决编译问题

**Pull request merged**